### PR TITLE
TINY-9829: Dispatch `insertFromPaste` input event upon pasting

### PR DESCRIPTION
**Pull request by mkzhx**
Related Ticket: TINY-9829

Description of Changes:
* Rename `FakeInputEvents` module to `InputEvents` and place it under `modules/tinymce/src/core/main/ts/events`
* Dispatch an `insertFromPaste` input event with relevant `data` property after a paste action (incl. image paste) occurs

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [x] Milestone set
* [x] ~Docs ticket created (if applicable)~

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/core/main/ts/keyboard/DeleteBackspaceKeys.ts [hunk 1]
```diff
@@ -44,11 +44,11 @@ const executeKeydownOverride = (editor: Editor, caret: Cell<Text | null>, evt: K
     .filter((_) => editor.selection.isEditable())
     .each((applyAction) => {
       evt.preventDefault();
-      const beforeInput = fireFakeBeforeInputEvent(editor, inputType);
+      const beforeInput = fireBeforeInputEvent(editor, inputType);
```
Comment (spocke) [inline modules/tinymce/src/core/main/ts/keyboard/DeleteBackspaceKeys.ts:47]: While you are at it maybe use namespace imports for this module everywhere it's being used.

File: modules/tinymce/src/core/main/ts/paste/Clipboard.ts [hunk 2]
```diff
@@ -30,9 +31,10 @@ const uniqueId = PasteUtils.createIdGenerator('mceclip');
 
 const doPaste = (editor: Editor, content: string, internal: boolean, pasteAsText: boolean): void => {
   const args = ProcessFilters.process(editor, content, internal);
-
   if (!args.cancelled) {
-    SmartPaste.insertContent(editor, args.content, pasteAsText);
+    const content = args.content;
+    SmartPaste.insertContent(editor, content, pasteAsText);
+    fireInputEvent(editor, 'insertFromPaste', { data: content });
```
Comment (spocke) [inline modules/tinymce/src/core/main/ts/paste/Clipboard.ts:37]: I don't think this event is correct. Check this codepen in Chrome https://codepen.io/spocke78/pen/PoydNMO it only populates the dataTransfer and only on beforeInput not on input. Also data never seems to be populated at least not on Chrome. So I think we should just mimic the browser here and not include the data. If people want to intercept paste data and alter it there are other ways. Like `BeforeSetContent` having a `paste` flag when stuff is coming from the clipboard.

File: modules/tinymce/src/core/test/ts/browser/paste/ImagePasteTest.ts [hunk 3]
```diff
@@ -5,23 +5,32 @@ import { TinyAssertions, TinyDom, TinyHooks, TinySelections } from '@ephox/wrap-
 import { assert } from 'chai';
 
 import Editor from 'tinymce/core/api/Editor';
+import { EditorEvent } from 'tinymce/core/api/util/EventDispatcher';
 import * as Clipboard from 'tinymce/core/paste/Clipboard';
 
 describe('browser.tinymce.core.paste.ImagePasteTest', () => {
+  let lastInputEvent: EditorEvent<InputEvent> | undefined;
```
Comment (spocke) [inline modules/tinymce/src/core/test/ts/browser/paste/ImagePasteTest.ts:12]: Use a `Singleton.value` wraps cell and option.

File: modules/tinymce/src/core/test/ts/browser/paste/InternalClipboardTest.ts [hunk 4]
```diff
@@ -9,10 +10,20 @@ import { EditorEvent } from 'tinymce/core/api/util/EventDispatcher';
 import * as InternalHtml from 'tinymce/core/paste/InternalHtml';
 import TablePlugin from 'tinymce/plugins/table/Plugin';
 
+interface ProcessEventExpectedData {
```
Comment (spocke) [inline modules/tinymce/src/core/test/ts/browser/paste/InternalClipboardTest.ts:13]: Maybe make these readonly unless they are meant to mutate things.

File: modules/tinymce/src/core/test/ts/browser/paste/InternalClipboardTest.ts [hunk 5]
```diff
@@ -9,10 +10,20 @@ import { EditorEvent } from 'tinymce/core/api/util/EventDispatcher';
...
+
+interface InputEventExpectedData {
+  data: string;
+}
+
 describe('browser.tinymce.core.paste.InternalClipboardTest', () => {
   let dataTransfer: DataTransfer | undefined;
   let lastPreProcessEvent: EditorEvent<PastePreProcessEvent> | undefined;
   let lastPostProcessEvent: EditorEvent<PastePostProcessEvent> | undefined;
+  let lastInputEvent: EditorEvent<InputEvent> | undefined;
```
Comment (spocke) [inline modules/tinymce/src/core/test/ts/browser/paste/InternalClipboardTest.ts:26]: These undefined things might predate katamari. I think it better to use singletons.

File: modules/tinymce/src/core/test/ts/browser/paste/PasteTest.ts [hunk 6]
```diff
@@ -148,7 +148,7 @@ describe('browser.tinymce.core.paste.PasteTest', () => {
     TinySelections.setSelection(editor, [ 0, 0 ], 1, [ 0, 0 ], 2);
 
     editor.execCommand('mceInsertClipboardContent', false, { text: ' a ' });
-    TinyAssertions.assertContent(editor, '<p>t&nbsp;a&nbsp;xt</p>');
+    TinyAssertions.assertContent(editor, '<p>t a xt</p>');
```
Comment (spocke) [inline modules/tinymce/src/core/test/ts/browser/paste/PasteTest.ts:151]: What happened here why is this now a space not a nbsp?

File: modules/tinymce/src/core/main/ts/paste/Clipboard.ts (hunk 2 above)
Comment (mkzhx) [reply modules/tinymce/src/core/main/ts/paste/Clipboard.ts:37]: @spocke Thanks for the demo! I see now that `data` is null across all browsers. 

However, both Firefox and Safari seems to have a populated dataTransfer property in the `input` event as well, but I'm happy to go with Chrome's approach of having both `data` and `dataTransfer` properties being `null`. 

Having said that, is it be worth also dispatching a `beforeInput` event as well which has a `DataTransfer` object containing the data, to further mimic browser behavior. If so, I will split that into a separate ticket.

File: modules/tinymce/src/core/test/ts/browser/paste/PasteTest.ts (hunk 6 above)
Comment (mkzhx) [reply modules/tinymce/src/core/test/ts/browser/paste/PasteTest.ts:151]: @spocke It seems to be editor behavior to convert any `&nbsp;` in the content to regular spaces after an `input` event is dispatched.
image

Comment (mkzhx) [reply modules/tinymce/src/core/test/ts/browser/paste/PasteTest.ts:151]: The line that normalizes `&nbsp;` on `input` is here: https://github.com/tinymce/tinymce/blob/b4c1a77689f2cd7cdd65bd4456f7de405430f658/modules/tinymce/src/core/main/ts/keyboard/InputKeys.ts#L8

File: modules/tinymce/src/core/test/ts/browser/paste/ImagePasteTest.ts (hunk 3 above)
Comment (mkzhx) [reply modules/tinymce/src/core/test/ts/browser/paste/ImagePasteTest.ts]: Good to know, will use `Singleton.value` in similar situations in the future.

**Review by MitchCTiny — APPROVED** [review]

File: modules/tinymce/src/core/test/ts/browser/paste/PasteTest.ts (hunk 6 above)
Comment (spocke) [reply modules/tinymce/src/core/test/ts/browser/paste/PasteTest.ts:154]: Right that makes sense now.

**Review by spocke — APPROVED** [review]

**Pull request merged**